text/0000-symbol-name-mangling-v2.md

```diff
@@ -682,42 +973,95 @@ Itanium mangling).
 # Unresolved questions
 [unresolved-questions]: #unresolved-questions
 
+### Complex Constant Data
+
+The RFC encodes constant values via the `<const-data> = {<hex-digit>} "_"`
+production, where `{<hex-digit>}` is the numeric value of the constant, not
+its representation as bytes. Using the numeric value is platform independent
+but does not easily scale to non-integer data types.
```

Comment:
I think we can use the byte representation for non-scalars, although that would not be platform-independent anymore.

The tricky bit is really pointers, which we could handle like miri does...

But if we want to only allow `#[structural_match]` types (originally meant for pattern-matching), then we can do something slightly better, which is represent them in the way a pattern would be.

IIRC, `miri` already has logic for lifting a value into a pattern, producing primitive leaves (integers, bools, chars and strings IIRC) and nested refutable/irrefutable constructors (e.g. tuples, structs, enum variants) - cc @oli-obk.

So we could define an extended const encoding, without losing compatibility with this RFC, which uses non-primitive types in `<const>` to indicate printing `TheType { 0: ..., 1: ..., ... }` (not sure we should bother with field names in the mangling), where `TheType` is a path so we can even use it to refer to a variant, for enums!

Sadly, the `<type>` inside the `<const>` is awkwardly placed whichever way we do it, unless we don't really use a `<type>` but something that can support nesting.

If we want to stay maximally compatible with any scheme for non-primitive-type const generics, maybe we should use `<basic-type>` instead of `<type>` in `<const>`?
My demangler implementation already gives up if it's not an unsigned integer type *anyway*.

Reply:
> IRC, miri already has logic for lifting a value into a pattern, producing primitive leaves (integers, bools, chars and strings IIRC) and nested refutable/irrefutable constructors (e.g. tuples, structs, enum variants)

Actually, the pattern matching code has that logic somewhat (it's limited on purpose). It's very problematic to just arbitrarily expand constants into patterns. E.g. if you have a byte literal read from a file, it could have thousands of bytes. expanding that into a slice pattern of byte patterns is very memory and cpu intensive.

Reply:
Aren't there byte literal patterns? Anyway, `str` and `[u8]` can be serialized as the bytes themselves (i.e. like a literal), no need to treat them like aggregates.

Reply:
In order to recognize that the `b"foo"` pattern is unreachable, we need to do some sort of deaggregation. At least in the way the current exhaustiveness checks are written.

```rust
    match b"foo" {
        b"bar" => {},
        [b'f', b'o', x] => println!("{}", x),
        b"foo" => {}, // unreachable pattern
        _ => {}
    }
```